Ruby: Add heredoc literals

sj26 opened the pull request:
This is a first pass at adding heredoc string literals for Ruby:

https://ruby-doc.org/core-2.7.0/doc/syntax/literals_rdoc.html#label-Here+Documents+-28heredocs-29

This only adds the basic form:

```ruby
<<STRING
foo #{42} bar
STRING
```

There are plenty of other styles with slightly different forms not yet covered by this PR:

```ruby
<<'STRING'
foo #{42} bar
STRING
```

```ruby
<<`STRING`
test 0 -eq #{0}
STRING
```

```ruby
def baz
  <<-STRING
  foo #{42} bar
  STRING
end
```

```ruby
def baz
  <<~STRING
    foo #{42} bar
  STRING
end
```

RunDevelopment commented on tests/run.js:
```diff
@@ -28,13 +28,15 @@ for (const language in testSuite) {
-						TestCase.runTestsWithHooks(language, require(filePath));
-					}
-				});
+				if (fileName[0] != '.') {
```
Please do not modify the test suite.

Why did you modify this in the first place?

sj26 replied:
Ah, sorry, I should have explained. I was editing one of the test files with vim, and the test runner was also trying to run my vim swap files, with errors like:

```
  1) Testing language 'ruby'
       – should pass test case '.string_feature.test':
     .../tests/languages/ruby/.string_feature.test.swp:1
b0VIM 8.2
      ^^^

SyntaxError: Unexpected number
      at wrapSafe (node:internal/modules/cjs/loader:1024:16)
      at Module._compile (node:internal/modules/cjs/loader:1072:27)
      at Object.Module._extensions..js (node:internal/modules/cjs/loader:1137:10)
      at Module.load (node:internal/modules/cjs/loader:973:32)
      at Function.Module._load (node:internal/modules/cjs/loader:813:14)
      at Module.require (node:internal/modules/cjs/loader:997:19)
      at require (node:internal/modules/cjs/helpers:92:18)
      at Context.<anonymous> (tests/run.js:35:44)
      at processImmediate (node:internal/timers:463:21)
```

So I made it skip hidden files, i.e. files starting with a dot (at least in *nix and macOS environments).

RunDevelopment replied:
I see. This is definitely a problem on our side.

I'll quickly fix the issue. Please revert this change and sync your fork.

RunDevelopment replied:
The problem has been fixed.

RunDevelopment commented on components/prism-ruby.js:
```diff
@@ -102,6 +102,21 @@
 			}
+		},
+		{
+			pattern: /<<([a-z_]\w*)[\r\n](?:.*[\r\n])*?\1/i,
```
Couldn't we easily add support for the other forms like this?

(The `delimiter` pattern has to be adjusted as well.)

```suggestion
			pattern: /<<[-~']?([a-z_]\w*)'?[\r\n](?:.*[\r\n])*?\1/i,
```

sj26 replied:
I was wondering about that. One of the forms disables interpolation, though:

```
<<'STRING'
will #{not} be interpolated
STRING
```

Can you vary whether interpolation is enabled by the contents of the match?

sj26 replied:
I was also hoping to add support for the heredoc anchor indicating the contents, like:

```
<<~SQL
  SELECT * FROM table WHERE column = 'value';
SQL
```

This is common in IDE syntax these days, and similar to what prism does for code fences in markdown which can contain other languages:

~~~
```sql
SELECT * FROM table WHERE column = 'value';
```
~~~

RunDevelopment replied:
> Can you vary whether interpolation is enabled by the contents of the match?

Yes and no. You can't change the `inside` tokens based on the match but you can just use two tokens like this:

```js
{
	pattern: /<<[-~`]?([a-z_]\w*)`?[\r\n](?:.*[\r\n])*?\1/i,
	alias: 'heredoc-string',
	greedy: true,
	inside: {
		'delimiter': {
			pattern: /^<<[a-z_]\w*|[a-z_]\w*$/i,
			alias: 'symbol',
			inside: {
				'punctuation': /^<<([-~]?)/
			}
		},
		'interpolation': interpolation
	}
},
{
	pattern: /<<'([a-z_]\w*)'[\r\n](?:.*[\r\n])*?\1/i,
	alias: 'heredoc-string',
	greedy: true,
	inside: {
		'delimiter': {
			pattern: /^<<[a-z_]\w*|[a-z_]\w*$/i,
			alias: 'symbol',
			inside: {
				'punctuation': /^<<([-~]?)/
			}
		}
	}
}
```

> I was also hoping to add support for the heredoc anchor indicating the contents

That's more difficult to do, especially when interpolation comes into play. 

Let's add support for heredoc first and do content highlighting later.

RunDevelopment reviewed (changes requested):
Thank you for the PR @sj26!

It looks mostly good as is. I left you a few comments and please commit the minified files as well (run `npm ci && npm run build` to generate those).

sj26 commented:
Thanks @RunDevelopment, I've removed the test suite change and expanded the heredoc support to more forms. It's not perfect, but it works well enough.

I wasn't able to run `npm ci && npm run build` for you though, sorry:

```
$ npm ci && npm run build

added 932 packages, and audited 933 packages in 7s

12 low severity vulnerabilities

To address issues that do not require attention, run:
  npm audit fix

To address all issues possible (including breaking changes), run:
  npm audit fix --force

Some issues need review, and may require choosing
a different dependency.

Run `npm audit` for details.

> prismjs@1.23.0 build
> gulp

[15:09:01] Using gulpfile ~/Developer/prism/gulpfile.js
[15:09:01] Starting 'default'...
[15:09:01] Starting 'minifyComponents'...
[15:09:01] Starting 'componentsJsonToJs'...
[15:09:01] Starting 'build'...
[15:09:01] Starting 'languagePlugins'...


#
# Fatal error in , line 0
# Check failed: allocator->SetPermissions(reinterpret_cast<void*>(region.begin()), region.size(), PageAllocator::kNoAccess).
#
#
#
#FailureMessage Object: 0x16dcb09d8
 1: 0x102249758 node::NodePlatform::GetStackTracePrinter()::$_3::__invoke() [/opt/nodenv/versions/15.6.0/bin/node]
 2: 0x102f91d50 V8_Fatal(char const*, ...) [/opt/nodenv/versions/15.6.0/bin/node]
 3: 0x1028e3b98 v8::internal::wasm::WasmCodeManager::Decommit(v8::base::AddressRegion) [/opt/nodenv/versions/15.6.0/bin/node]
 4: 0x1028e7218 v8::internal::wasm::NativeModule::FreeCode(v8::internal::Vector<v8::internal::wasm::WasmCode* const>) [/opt/nodenv/versions/15.6.0/bin/node]
 5: 0x1028f6400 v8::internal::wasm::WasmEngine::FreeDeadCodeLocked(std::__1::unordered_map<v8::internal::wasm::NativeModule*, std::__1::vector<v8::internal::wasm::WasmCode*, std::__1::allocator<v8::internal::wasm::WasmCode*> >, std::__1::hash<v8::internal::wasm::NativeModule*>, std::__1::equal_to<v8::internal::wasm::NativeModule*>, std::__1::allocator<std::__1::pair<v8::internal::wasm::NativeModule* const, std::__1::vector<v8::internal::wasm::WasmCode*, std::__1::allocator<v8::internal::wasm::WasmCode*> > > > > const&) [/opt/nodenv/versions/15.6.0/bin/node]
 6: 0x1028f4108 v8::internal::wasm::WasmEngine::PotentiallyFinishCurrentGC() [/opt/nodenv/versions/15.6.0/bin/node]
 7: 0x1028f557c v8::internal::wasm::WasmEngine::ReportLiveCodeForGC(v8::internal::Isolate*, v8::internal::Vector<v8::internal::wasm::WasmCode*>) [/opt/nodenv/versions/15.6.0/bin/node]
 8: 0x1028f592c v8::internal::wasm::WasmEngine::ReportLiveCodeFromStackForGC(v8::internal::Isolate*) [/opt/nodenv/versions/15.6.0/bin/node]
 9: 0x10247f348 v8::internal::StackGuard::HandleInterrupts() [/opt/nodenv/versions/15.6.0/bin/node]
10: 0x1027ba830 v8::internal::Runtime_StackGuard(int, unsigned long*, v8::internal::Isolate*) [/opt/nodenv/versions/15.6.0/bin/node]
11: 0x102aabb6c Builtins_CEntry_Return1_DontSaveFPRegs_ArgvOnStack_NoBuiltinExit [/opt/nodenv/versions/15.6.0/bin/node]
12: 0x102a44910 Builtins_InterpreterEntryTrampoline [/opt/nodenv/versions/15.6.0/bin/node]
13: 0x102b1fbf8 Builtins_SortCompareUserFn [/opt/nodenv/versions/15.6.0/bin/node]
14: 0x102b222dc Builtins_ArrayTimSort [/opt/nodenv/versions/15.6.0/bin/node]
15: 0x102b234c8 Builtins_ArrayPrototypeSort [/opt/nodenv/versions/15.6.0/bin/node]
16: 0x102a44874 Builtins_InterpreterEntryTrampoline [/opt/nodenv/versions/15.6.0/bin/node]
17: 0x102a44874 Builtins_InterpreterEntryTrampoline [/opt/nodenv/versions/15.6.0/bin/node]
18: 0x102a44874 Builtins_InterpreterEntryTrampoline [/opt/nodenv/versions/15.6.0/bin/node]
19: 0x102a44874 Builtins_InterpreterEntryTrampoline [/opt/nodenv/versions/15.6.0/bin/node]
20: 0x102a44874 Builtins_InterpreterEntryTrampoline [/opt/nodenv/versions/15.6.0/bin/node]
21: 0x102a3bea4 Builtins_ArgumentsAdaptorTrampoline [/opt/nodenv/versions/15.6.0/bin/node]
22: 0x102a44874 Builtins_InterpreterEntryTrampoline [/opt/nodenv/versions/15.6.0/bin/node]
23: 0x102a44874 Builtins_InterpreterEntryTrampoline [/opt/nodenv/versions/15.6.0/bin/node]
24: 0x102a44874 Builtins_InterpreterEntryTrampoline [/opt/nodenv/versions/15.6.0/bin/node]
25: 0x102a44874 Builtins_InterpreterEntryTrampoline [/opt/nodenv/versions/15.6.0/bin/node]
26: 0x102a3bea4 Builtins_ArgumentsAdaptorTrampoline [/opt/nodenv/versions/15.6.0/bin/node]
27: 0x102a44874 Builtins_InterpreterEntryTrampoline [/opt/nodenv/versions/15.6.0/bin/node]
28: 0x102a3bea4 Builtins_ArgumentsAdaptorTrampoline [/opt/nodenv/versions/15.6.0/bin/node]
29: 0x102a44874 Builtins_InterpreterEntryTrampoline [/opt/nodenv/versions/15.6.0/bin/node]
30: 0x102a44874 Builtins_InterpreterEntryTrampoline [/opt/nodenv/versions/15.6.0/bin/node]
31: 0x102a3bea4 Builtins_ArgumentsAdaptorTrampoline [/opt/nodenv/versions/15.6.0/bin/node]
32: 0x102a42124 Builtins_JSEntryTrampoline [/opt/nodenv/versions/15.6.0/bin/node]
33: 0x102a41dc8 Builtins_JSEntry [/opt/nodenv/versions/15.6.0/bin/node]
34: 0x158008000 
zsh: trace trap  npm run build
```

RunDevelopment commented:
> Maybe because I'm on an M1 mac?

Yes but don't worry. According to nodejs/node#37061, this issue has been fixed in v15.9.0. You just need to update your NodeJS version.

RunDevelopment commented on components/prism-ruby.js:
```diff
@@ -102,6 +102,35 @@
+			}
+		},
+		{
+			pattern: /<<[-~]?'([a-z_]\w*)'[\r\n](?:.*[\r\n])*?\s*\1/i,
```
Is `<<~'foo'` also allowed?

sj26 replied:
Yeah:

<img width="299" alt="image" src="https://user-images.githubusercontent.com/14028/117237671-c504e500-ae6e-11eb-9d65-006aeb10e865.png">

RunDevelopment commented:
I have no idea why the tests fail. Everything looks fine to me. I'll investigate this tomorrow.

RunDevelopment commented on components/prism-ruby.js:
```diff
@@ -102,6 +102,35 @@
+			}
+		},
+		{
+			pattern: /<<[-~]?'([a-z_]\w*)'[\r\n](?:.*[\r\n])*?\s*\1/i,
```
```suggestion
			pattern: /<<[-~]?'([a-z_]\w*)'[\r\n](?:.*[\r\n])*?[\t ]*\1/i,
```

sj26 commented:
Thanks @RunDevelopment, I've folded those changes in :+1:

RunDevelopment commented:
Please rebuild @sj26 and then we can merge this.

sj26 commented:
Sorry I forgot to rebuild, but done now :+1:

RunDevelopment commented:
Thank you for contributing @sj26!
